django/db/models/fields/composite.py

```diff
@@ -0,0 +1,192 @@
+from django.db.models.expressions import Col, ExpressionList
+from django.db.models.fields import Field
+from django.db.models.lookups import TupleExact
+from django.db.models.query_utils import DeferredAttribute
+from django.db.models.signals import class_prepared
+from django.utils.functional import cached_property
+
+
+class CompositeCol(ExpressionList):
+    def __init__(self, alias, target: "CompositeField", output_field=None):
+        if output_field is None:
+            output_field = target
+        self.cols = [Col(alias, field) for field in target.component_fields]
+        super().__init__(*self.cols, output_field=output_field)
+        self.alias, self.target = alias, target
+
+    def as_sql(self, compiler, connection):
+        sql, params = super().as_sql(compiler, connection)
+        return "(%s)" % sql, params
+
+    def get_lookup(self, lookup):
+        if lookup == "exact":
+            return TupleExact
+        return super().get_lookup(lookup)
+
+
+class CompositeType:
+    _composite_type = None
+    _instance = None
+    _repr_fmt = "()"
+    _fields = ()
+
+    @classmethod
+    def namedtuple(cls, name, fields):
+        fields = tuple(fields)
+        doc_fields = ",".join(fields)
+        repr_fields = ", ".join(f"{name}=%r" for name in fields)
+        class_namespace = {
+            "__doc__": f"{name}({doc_fields})",
+            "_repr_fmt": "({repr_fields})",
+            "__slots__": ("_instance",),
+            "_fields": fields,
+        }
+
+        def bind_property(name):
+            def fget(self):
+                return getattr(self._instance, name)
+
+            def fset(self, value):
+                return setattr(self._instance, name, value)
+
+            return property(fget, fset, doc=f"Bind for field {name}")
+
+        for field in fields:
+            class_namespace[field] = bind_property(field)
+        return type(name, (cls,), class_namespace)
+
+    def resolve_expression(self, *args, **kwargs):
+        def resolve_value(value):
+            if isinstance(value, CompositeType):
+                value = tuple(value)
+            if hasattr(value, "resolve_expression"):
+                value = value.resolve_expression(*args, **kwargs)
+            elif isinstance(value, (list, tuple)):
+                # The items of the iterable may be expressions and therefore need
+                # to be resolved independently.
+                values = (resolve_value(sub_value) for sub_value in value)
+                type_ = type(value)
+                if hasattr(type_, "_make"):  # namedtuple
+                    return type_(*values)
+                return type_(values)
```

Comment:
Could this be:

```suggestion
                values = (resolve_value(sub_value) for sub_value in value)
                if isinstance(value, namedtuple):
                    return namedtuple(*values)
                return type(value)(values)
```

I'm also wondering if we even need to maintain the type of `value` here - would it work to always cast to `tuple` or `list`?

```suggestion
                return [resolve_value(sub_value) for sub_value in value]
```